dlt/destinations/snowflake/snowflake.py

```diff
@@ -0,0 +1,164 @@
+from pathlib import Path
+from typing import ClassVar, Dict, Optional, Sequence, Tuple, List, cast, Iterable
+
+from dlt.common import json, logger
+from dlt.common.arithmetics import DEFAULT_NUMERIC_PRECISION, DEFAULT_NUMERIC_SCALE
+from dlt.common.configuration.specs import GcpServiceAccountCredentialsWithoutDefaults
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import FollowupJob, NewLoadJob, TLoadJobState, LoadJob
+from dlt.common.data_types import TDataType
+from dlt.common.storages.file_storage import FileStorage
+from dlt.common.schema import TColumnSchema, Schema, TTableSchemaColumns, TSchemaTables
+from dlt.common.schema.typing import TTableSchema, TWriteDisposition
+from dlt.common.wei import EVM_DECIMAL_PRECISION
+
+from dlt.destinations.job_client_impl import SqlJobClientBase
+from dlt.destinations.job_impl import EmptyLoadJob
+from dlt.destinations.exceptions import DestinationSchemaWillNotUpdate, DestinationTransientException, LoadJobNotExistsException, LoadJobTerminalException, LoadJobUnknownTableException
+
+from dlt.destinations.snowflake import capabilities
+from dlt.destinations.snowflake.configuration import SnowflakeClientConfiguration
+from dlt.destinations.snowflake.sql_client import SnowflakeSqlClient
+from dlt.destinations.sql_merge_job import SqlMergeJob
+from dlt.destinations.snowflake.sql_client import SnowflakeSqlClient
+
+BIGINT_PRECISION = 19
+MAX_NUMERIC_PRECISION = 38
+
+
+SCT_TO_BQT: Dict[TDataType, str] = {
+    "complex": "VARIANT",
+    "text": "VARCHAR",
+    "double": "FLOAT",
+    "bool": "BOOLEAN",
+    "date": "DATE",
+    "timestamp": "TIMESTAMP_TZ",
+    "bigint": f"NUMBER({BIGINT_PRECISION},0)",  # Snowflake has no integer types
+    "binary": "BINARY",
+    "decimal": f"NUMBER({DEFAULT_NUMERIC_PRECISION},{DEFAULT_NUMERIC_SCALE})",
+}
+
+BQT_TO_SCT: Dict[str, TDataType] = {
+    "VARCHAR": "text",
+    "FLOAT": "double",
+    "BOOLEAN": "bool",
+    "DATE": "date",
+    "TIMESTAMP_TZ": "timestamp",
+    "BINARY": "binary",
+    "VARIANT": "complex"
+}
+
+class SnowflakeLoadJob(LoadJob, FollowupJob):
+    def __init__(self, file_path: str, table_name: str, write_disposition: TWriteDisposition, load_id: str, client: SnowflakeSqlClient) -> None:
+        file_name = FileStorage.get_file_name_from_file_path(file_path)
+        super().__init__(file_name)
+
+        with client.with_staging_dataset(write_disposition == "merge"):
+            qualified_table_name = client.make_qualified_table_name(table_name)
+            stage_name = f'"{table_name}_{load_id}"'
+            # Create stage and file format
+            client.execute_sql(
+                """
+                CREATE FILE FORMAT IF NOT EXISTS dlt_jsonl_format
+                TYPE = 'JSON'
+                BINARY_FORMAT = 'BASE64'
+                """
+            )
+            client.execute_sql(f"CREATE STAGE IF NOT EXISTS {stage_name} FILE_FORMAT = dlt_jsonl_format")
```

Comment:
please do not create a stage. use the table implicit stage: https://docs.snowflake.com/en/user-guide/data-load-local-file-system-create-stage

use the `load_id` as the folder in stage from which you will copy

let the user specify a stage name in the configuration. if the stage name is available, this stage will be used

Reply:
Updated default stage  https://github.com/dlt-hub/dlt/pull/414/commits/bf322ecd8a972290e7c28d652244c585ef30df17#diff-577441894b0e65b848310ea282c52bab00b6f649eb2c1b9402a5423d70e4d3e4

Reply:
Added the config option to use a different stage, it gets created if it doesn't exist. And optionally delete the stage file after copy: https://github.com/dlt-hub/dlt/pull/414/commits/6853c2747e5b5bac1c2a77233292ef81a9a541bc